File: Python/BuilderPattern.py
# Base class representing a generic car
class Car:
    def __init__(self):
        self.make = None
        self.model = None
        self.color = None

    def __str__(self):
        return f"Make: {self.make}, Model: {self.model}, Color: {self.color}"


# Abstract builder class for constructing a car
class CarBuilder:
    def __init__(self):
        self.car = Car()

    def set_make(self, make):
        self.car.make = make
        return self

    def set_model(self, model):
        self.car.model = model
        return self

    def set_color(self, color):
        self.car.color = color
        return self

    def build(self):
        return self.car


# Concrete builder class for constructing a manual transmission car
class ManualCarBuilder(CarBuilder):
    def __init__(self):
        super().__init__()
        self.car.transmission = "Manual"  # Set default transmission

    def set_gear(self, gear):
        self.car.gear = gear
        return self


# Concrete builder class for constructing an automatic transmission car
class AutomaticCarBuilder(CarBuilder):
    def __init__(self):
        super().__init__()
        self.car.transmission = "Automatic"  # Set default transmission

    def set_drive_mode(self, mode):
        self.car.drive_mode = mode
        return self


# Concrete builder class for constructing an electric car
class ElectricCarBuilder(CarBuilder):
    def __init__(self):
        super().__init__()
        self.car.engine_type = "Electric"  # Set default engine type

    def set_battery_capacity(self, capacity):
        self.car.battery_capacity = capacity
        return self


# Director class responsible for orchestrating the construction process
class CarDirector:
    def __init__(self, builder):
        self.builder = builder

    def construct_car(self, make, model, color):
        return self.builder.set_make(make)\
                           .set_model(model)\
                           .set_color(color)\
                           .build()


# Usage
if __name__ == "__main__":
    # Create instances of different builders
    manual_builder = ManualCarBuilder()
    automatic_builder = AutomaticCarBuilder()
    electric_builder = ElectricCarBuilder()

    # Construct manual car
    director = CarDirector(manual_builder)
    manual_car = director.construct_car("Toyota", "Corolla", "Red")

    # Construct automatic car
    director = CarDirector(automatic_builder)
    automatic_car = director.construct_car("Honda", "Civic", "Blue")

    # Construct electric car
    director = CarDirector(electric_builder)
    electric_car = director.construct_car("Tesla", "Model S", "White")

    # Print details of constructed cars
    print(manual_car)
    print(automatic_car)
    print(electric_car)

File: Javascript/BuilderPattern.js
// Base class representing a generic car
class Car {
  constructor() {
    this.make = null;
    this.model = null;
    this.color = null;
    this.transmission = null; // Transmission type
    this.batteryType = null; // Engine type
  }

  toString() {
    return `Make: ${this.make}, Model: ${this.model}, Color: ${this.color}, Transmission: ${this.transmission}, Battery Type: ${this.batteryType}`;
  }
}

// Abstract builder class for constructing a car
class CarBuilder {
  constructor() {
    this.car = new Car();
  }

  setMake(make) {
    this.car.make = make;
    return this;
  }

  setModel(model) {
    this.car.model = model;
    return this;
  }

  setColor(color) {
    this.car.color = color;
    return this;
  }

  build() {
    return this.car;
  }
}

// Concrete builder class for constructing a manual transmission car
class ManualCarBuilder extends CarBuilder {
  constructor() {
    super();
    this.car.transmission = "Manual";
  }
}

// Concrete builder class for constructing an automatic transmission car
class AutomaticCarBuilder extends CarBuilder {
  constructor() {
    super();
    this.car.transmission = "Automatic";
  }
}

// Concrete builder class for constructing an electric car
class ElectricCarBuilder extends CarBuilder {
  constructor() {
    super();
    this.car.batteryType = "Electric";
  }
}

// Director class responsible for orchestrating the construction process
class CarDirector {
  constructor(builder) {
    this.builder = builder;
  }

  constructCar(make, model, color) {
    return this.builder.setMake(make).setModel(model).setColor(color).build();
  }
}

// Usage
const manualBuilder = new ManualCarBuilder();
const automaticBuilder = new AutomaticCarBuilder();
const electricBuilder = new ElectricCarBuilder();

const director1 = new CarDirector(manualBuilder);
const manualCar = director1.constructCar("Toyota", "Corolla", "Red");

const director2 = new CarDirector(automaticBuilder);
const automaticCar = director2.constructCar("Honda", "Civic", "Blue");

const director3 = new CarDirector(electricBuilder);
const electricCar = director3.constructCar("Tesla", "Model S", "White");

console.log(manualCar.toString());
console.log(automaticCar.toString());
console.log(electricCar.toString());
